text/0000-union-initialization-and-drop.md

```diff
@@ -0,0 +1,367 @@
+variable has to be dropped when it goes out of scope.
+
+A union just does very simple initialization tracking: There is a single boolean
+state for the entire union and all of its fields.  Nested inner fields are
+tracked just like they are for structs; however, when the union becomes
+(un)initialized, then all nested inner fields of all union fields are
+(un)initialized at once.  So, (un)initializing a union field also
+(un)initializes its siblings.  For example:
+
+```rust
+// This code creates bad references and transmutes to `Vec` in incorrect ways.
+// This is just to demonstrate what the compiler would accept in terms of
+// tracking initialization.
+
+struct S(i32); // not `Copy`, no drop glue
+union U { f1: ManuallyDrop<Vec<i32>>, f2: (S, S), f3: i32 }
+
+let mut u: U;
+// Now `u` is not initialized: `&u`, `&u.f2` and `&u.f2.0` are all rejected.
+
+// We can write into uninitialized inner fields:
+u.f2.1 = S(42);
+{ let _x = &u.f2.1; } // This field is initialized now.
+// But this does not change the initialization state of the union itself,
+// or any other (inner) field.
+
+// We can initialize by assigning an entire field:
+u.f1 = ManuallyDrop::new(Vec::new());
+// Now *all (nested) fields* of `u` are initialized, including the siblings of `f1`:
+{ let _x = &u.f2; }
+{ let _x = &u.f2.0; }
+
+// Equivalently, we can assign the entire union:
+u = U { f2: (S(42), S(23) };
+// Now `u` is still initialized.
+
+// Copying does not change anything:
+let _x = u.f3;
+// Now `u` is still initialized.
+
+// We can move out of an initialized union:
+let v = u.f1;
+// Now `f1` *and its siblings* are no longer initialized (they got "moved out of"):
+// `let _x = u.f2;` would hence get rejected, as would `&u.f1` and `foo(u)`.
+u.f1 = v;
+// Now `u` and all of its fields are initialized again ("moving back in").
+
+// When we move out of an inner field, the other union fields become uninitialized
+// even if they are `Copy`.
+let s = u.f2.1;
+// Now `u.f1` and `u.f3` are no longer initialized.  But `u.f2.0` is:
+let s = u.f2.0;
+```
+
+If the union implements `Drop`, the same restrictions as for structs apply: It
+is not possible to initialize a field before initializing the entire variable,
+and it is not possible to move out of a field.  For example:
+
+```rust
+// This code creates bad references and transmutes to `Vec` in incorrect ways.
+// This is just to demonstrate what the compiler would accept in terms of
+// tracking initialization.
+
+struct S(i32); // not `Copy`, no drop glue
+
+union U { f1: ManuallyDrop<Vec<i32>>, f2: (S, S), f3: u32 }
+impl Drop for U {
+    fn drop(&mut self) {
+        println!("Goodbye!");
+    }
+}
+
+let mut u: U;
+// `u.f1 = ...;` gets rejected: Cannot initialize a union with `Drop` by assigning a field.
+u = U { f2: (S(42), S(1)) };
+// Now `u` is initialized.
+
+// `let v = u.f1;` gets rejected: Cannot move out of union that implements `Drop`.
```

Comment:
If we can't move out of `Drop` unions, how would I change which field of the union is initialized without losing access to the previous value?

Reply:
How would I express this sort of thing?
```rust
union U { f1: ManuallyDrop<String>, f2: u64 }
impl Drop for U { ... }
let mut x = U { f1: ManuallyDrop::new("foo".to_string()) };
...
let my_string = x.f1;
x.f2 = 5;
...
```

Reply:
My gut feeling is, if you want that -- don't implement `Drop` for your union. Add a wrapper struct around it if you need it for the public interface.

We could also say that we allow moving out of dropping unions. Because moving out will deinitialize all fields, unlike for structs, this cannot lead to a "half initialized" solution. So after just
```
union U { f1: ManuallyDrop<String>, f2: u64 }
impl Drop for U { ... }
let mut x = U { f1: ManuallyDrop::new("foo".to_string()) };
...
let my_string = x.f1;
```
`drop` would NOT be called, but if you then do `x.f2 = 5`, it would again be called.

Seems easy to miss though, so I wanted to be conservative.

Reply:
I'd be in favor of allowing moving out and having that prevent `Drop`, but I don't feel super strongly and you're right that any use-case I can come up with would be easily solved via a wrapper struct. i agree that it's better to be conservative here.

Reply:
I remember there was an RFC about moving out from `Drop` structs last year (?) that was postponed, but can't find it right now.
If one field is moved out, then other fields would still be dropped, but the struct itself wouldn't.
Same rule is applicable for unions as well, except that "other fields would still be dropped" would be no-op.

Reply:
I added this as an unresolved question.